### ISSUE TITLE
[browser][MT] Multithreading and JavaScript async interop in .NET 9

### ISSUE THREAD

**Issue by lambdageek** [dotnet/runtime#85592]
Tracking issue for further work on JS interop and multithreading.

Constituent part of https://github.com/dotnet/runtime/issues/76956

## Goals
- CPU intensive workloads on dotnet thread pool
- Allow user to start new managed threads using `new Thread` and join it.
- Add new C# API for creating dedicated web workers with JS interop. Allow JS async/promises via external event loop.
- enable blocking `Task.Wait` and `lock()` like APIs from C# user code on all threads or non-JS/UI threads
    - Current public API throws PNSE for it
    - This is core part on MT value proposition.
    - If people want to use existing MT code-bases, most of the time, the code is full of locks.
    - People want to use existing desktop/server multi-threaded code as is.
- allow HTTP and WS C# APIs to be used from any thread despite underlying JS object affinity
- JSImport/JSExport interop in maximum possible extent.
- don't change/break single threaded build. †

## Lower priority goals
- try to make it debugging friendly
- sync C# to async JS
    - dynamic creation of new pthread
    - implement crypto via `subtle` browser API
    - allow MonoVM to lazily download DLLs from the server, instead of during startup.
    - implement synchronous APIs of the HTTP and WS clients. At the moment they throw PNSE.
- sync JS to async JS to sync C#
    - allow calls to synchronous JSExport from UI thread (callback)
- don't prevent future marshaling of JS [transferable objects](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects), like streams and canvas.
- offload CPU intensive part of WASM startup to WebWorker, so that the pre-rendered (blazor) UI could stay responsive during Mono VM startup.

## Non-goals
- interact with JS state on WebWorker of managed threads other than UI thread or dedicated `JSWebWorker`

## Design discussion and experiment
* [x] Design discussion https://github.com/dotnet/designs/pull/301
* [x] test deputy/sidecar design with Blazor (in some limited way)
* https://github.com/dotnet/runtime/pull/92226
* https://github.com/dotnet/runtime/pull/93920
* https://github.com/dotnet/runtime/pull/94525

## Depending on selected design
* [x] call dispatch to/from UI thread
* [x] call dispatch to/from JSWebWorker thread
* [x] Make Promise -> Task marshaling ahead of call
* [x] Make string marshaling non GC boundary breaching
* [x] Make Array marshaling non GC boundary breaching
* [x] dispatch calls with JS affinity to correct thread. Don't make the user responsible for it.
* [x] Dynamic allocation of managed threads
    * `PTHREAD_POOL_SIZE_STRICT=0`
    * https://github.com/dotnet/runtime/issues/91538
    * https://github.com/dotnet/runtime/issues/75629
* [x] schedule disposal of `JSObject` proxy on correct thread later, so that it doesn't deadlock GC by dispatching to suspended thread
* [x] Make HTTP/WS clients dispatch more async
* [x] Solve sync dispatch to HTTP/WS
* [x] Blazor's `WebAssemblyDispatcher` more async, learn from feedback https://github.com/dotnet/aspnetcore/pull/48991
* [x] review Blazor legacy interop and JS interop
* [x] https://github.com/dotnet/runtime/issues/97085

## Bugs
* https://github.com/dotnet/runtime/issues/90044
* https://github.com/dotnet/runtime/issues/86903
* https://github.com/dotnet/runtime/issues/88663
* https://github.com/dotnet/runtime/issues/97546
* https://github.com/dotnet/runtime/issues/96190
* https://github.com/dotnet/runtime/issues/98406
* https://github.com/dotnet/runtime/issues/97913
* https://github.com/dotnet/runtime/issues/98771
* https://github.com/dotnet/runtime/issues/97652

## Memory growth, alignment
* https://github.com/dotnet/runtime/issues/96545
* https://github.com/dotnet/runtime/issues/98630
* https://github.com/dotnet/runtime/issues/98217
* https://github.com/dotnet/runtime/issues/96191
* https://github.com/dotnet/runtime/issues/96546

## Broken tests
* https://github.com/dotnet/runtime/issues/75132
* https://github.com/dotnet/runtime/issues/93323
* https://github.com/dotnet/runtime/issues/91541
* https://github.com/dotnet/runtime/issues/75123
* https://github.com/dotnet/runtime/issues/91661
* https://github.com/dotnet/runtime/issues/91597
* https://github.com/dotnet/runtime/issues/91550
* https://github.com/dotnet/runtime/issues/91552
* https://github.com/dotnet/runtime/issues/91577
* https://github.com/dotnet/runtime/issues/91581
* https://github.com/dotnet/runtime/issues/91579
* https://github.com/dotnet/runtime/issues/91582
* https://github.com/dotnet/runtime/issues/91547
* https://github.com/dotnet/runtime/issues/94486
* https://github.com/dotnet/runtime/issues/91583
* https://github.com/dotnet/runtime/issues/91676
* https://github.com/dotnet/runtime/issues/95795
* https://github.com/dotnet/runtime/issues/96173
* https://github.com/dotnet/runtime/issues/96191
* https://github.com/dotnet/runtime/issues/97513
* https://github.com/dotnet/runtime/issues/97382
* https://github.com/dotnet/runtime/issues/97092
* https://github.com/dotnet/runtime/issues/96628
* https://github.com/dotnet/runtime/issues/98110
* https://github.com/dotnet/runtime/issues/98101
* https://github.com/dotnet/runtime/issues/98035
* https://github.com/dotnet/runtime/issues/98026
* https://github.com/dotnet/runtime/issues/97968
* https://github.com/dotnet/runtime/issues/97914
* https://github.com/dotnet/runtime/issues/98216
* https://github.com/dotnet/runtime/issues/98219
* https://github.com/dotnet/runtime/issues/98344
* https://github.com/dotnet/runtime/issues/99519
* https://github.com/dotnet/runtime/issues/44622
* https://github.com/dotnet/runtime/issues/99500
* https://github.com/dotnet/runtime/issues/99499
* https://github.com/dotnet/runtime/issues/99501
* https://github.com/dotnet/runtime/issues/99951
* https://github.com/dotnet/runtime/issues/100275
* https://github.com/dotnet/runtime/issues/100249

## Nice to have
* [ ] more scenarios on Radek's perf bench
* [x] Stop using ThreadlessXunitTestRunner https://github.com/dotnet/runtime/issues/88763
* [x] re-consider counting unsettled promises https://github.com/dotnet/runtime/issues/84489
* [x] re-consider pending user `setTimeout` https://github.com/dotnet/runtime/issues/85052
* [ ] noisy emscripten https://github.com/dotnet/runtime/issues/75625
* [x] throw on blocking calls from UI https://github.com/dotnet/runtime/issues/76958
* [ ] blazor background startup https://github.com/dotnet/runtime/issues/65559
* [x] console perf https://github.com/dotnet/runtime/issues/93603
* [x] shutdown all threads/workers during `exit()` and `abort()`
* [x] synchronous JS APIs on the UI thread and spin-wait
* [ ] other sync to async scenarios see above
* [x] pthread pool size discussion https://github.com/dotnet/runtime/issues/75602
* [ ] https://github.com/dotnet/runtime/issues/98214
* [x] https://github.com/dotnet/runtime/issues/98365
* [x] https://github.com/dotnet/runtime/issues/98309
* [ ] https://github.com/dotnet/runtime/issues/100273

## Progress
* https://github.com/dotnet/runtime/pull/94217
* https://github.com/dotnet/runtime/pull/94214
* https://github.com/dotnet/runtime/pull/94204
* https://github.com/dotnet/runtime/pull/94180
* https://github.com/dotnet/runtime/pull/93648
* https://github.com/dotnet/runtime/pull/94466
* https://github.com/dotnet/runtime/pull/95177
* https://github.com/dotnet/runtime/pull/95180
* https://github.com/dotnet/runtime/pull/95411
* https://github.com/dotnet/runtime/pull/95483
* https://github.com/dotnet/runtime/pull/95783
* https://github.com/dotnet/runtime/pull/95702
* https://github.com/dotnet/runtime/pull/96319
* https://github.com/dotnet/runtime/pull/96504
* https://github.com/dotnet/aspnetcore/pull/52724
* https://github.com/dotnet/runtime/pull/97669
* https://github.com/dotnet/runtime/pull/98298
* https://github.com/dotnet/runtime/pull/98391
* https://github.com/dotnet/runtime/pull/99524
* https://github.com/dotnet/runtime/pull/99422
* https://github.com/dotnet/runtime/pull/99833
* https://github.com/dotnet/runtime/pull/99296
* https://github.com/dotnet/runtime/pull/99347
* https://github.com/dotnet/runtime/pull/99397
* https://github.com/dotnet/runtime/pull/100415
* https://github.com/dotnet/runtime/pull/100778
* https://github.com/dotnet/runtime/pull/100610

## Future
* API for `JSWebWorker` with external loop and JS interop
* `JSSynchronizationContext` for UI and `JSWebWorker` threads
* `JSSynchronizationContext` for `JSObject` proxy
* `[UnsupportedOSPlatform]` ref assemblies https://github.com/dotnet/runtime/issues/88147
* https://github.com/dotnet/runtime/issues/76957
* https://github.com/dotnet/runtime/issues/76963
* https://github.com/dotnet/runtime/issues/100411
* https://github.com/dotnet/runtime/issues/98366
* https://github.com/dotnet/runtime/issues/92853 
* https://github.com/dotnet/runtime/issues/87556
* https://github.com/dotnet/runtime/issues/76964

**Comment by Rippletank** [comment]
I'm using the latest  .net8.0 RC2 release in a multithreaded WASM app using UNO and Skiasharp. I am getting the "Please use dedicated worker...etc" from the AssertWebWorkerContext() method as expected.

The work around is sending JSInterop calls to the main thread, which works but results in some animation stuttering particularly with large db/cryptography jobs. 

I would like to try out the WebWorker class but it does not appear to be included in the Microsoft.NETCore.App\8.0.0-rc.2.23479.6\System.Runtime.InteropServices.JavaScript.dll which the application is using and because of the internal methods I see how to implement it outside of the dotnet runtime.

Is this just not available in released packages yet or is there some way to enable it?

I notice there is a [sample](https://github.com/dotnet/runtime/tree/e4d47362b1b8f31f812db45e037553c913141469/src/mono/sample/wasm/browser-threads-minimal) that uses it but doesn't this need compiling the runtime from source?

**Comment by pavelsavara** [comment]
> I'm using the latest .net8.0 RC2 release in a multithreaded WASM app using UNO and Skiasharp. 

@Rippletank Threads are not supported in .Net 8. As you can see above, there are many reasons why we are not ready for it.

**Comment by jtorjo** [comment]
@pavelsavara Thanks for the update. I'm curious if we'll ever have this, because if I remember correctly, this was .net7 --> .net8, and now it's .net8 --> .net9

**Comment by Rippletank** [comment]
> @Rippletank Threads are not supported in .Net 8. As you can see above, there are many reasons why we are not ready for it.

Ok, I see. It is confusing because multithreading is enabled with Uno web assembly apps and it works within the app itself but obviously doesn't work with the interop. 

I'm curious, is ok to use it, apart from interop issues, or is it generally problematic at the moment? For things like http etc.

**Comment by pavelsavara** [comment]
> I'm curious, is ok to use it, apart from interop issues, or is it generally problematic at the moment? For things like http etc.

MT is problematic in Net8 in my experience, that's why it's experimental. I suggest you ask Uno how to turn it of.

> @pavelsavara Thanks for the update. I'm curious if we'll ever have this, because if I remember correctly, this was .net7 --> .net8, and now it's .net8 --> .net9

Yes those are not easy problems. We are not out of the woods yet for Net9 either. We don't want to ship product which could randomly deadlock. Wish us luck.

**Comment by jtorjo** [comment]
> Yes those are not easy problems. We are not out of the woods yet for Net9 either. We don't want to ship product which could randomly deadlock. Wish us luck.

@pavelsavara  I'm definitely wishing you luck. But it seems we won't have it for another 2+ years, if I understand you correctly.

**Comment by pavelsavara** [comment]
> @pavelsavara I'm definitely wishing you luck. But it seems we won't have it for another 2+ years, if I understand you correctly.

I understand that you are waiting with some specific use-case for this ? Could you please share more details why you would benefit from threads in browser ? Also anyone else reading this, I would like to hear about your use-cases.

**Comment by jtorjo** [comment]
@pavelsavara I work at [OpenSilver](https://opensilver.net/), and we're migrating Silverlight apps to OpenSilver (which ends up compiling to webassembly).

Thus, we end up going from apps that were written using multiple threads (Silverlight), to a single thread (webassembly). Most of the time, this is not an issue, but every now we end up with some bottlenecks that are insanely hard to fix (time-consuming parts of the code will end up being executed, freezing the UI).

That's where multi-threading would help.

**Comment by ivanjx** [comment]
@pavelsavara i would like to do image processing (like qr/barcode scanning) in separate threads.

**Comment by Rippletank** [comment]
Similar case here. Our educational app was built in Xamarin, now Maui but really it's a SkiaSharp app with a few support services supplied by the host. Getting it running as an Uno WebAssembly app was very quick but its taken much much longer to get it running close to acceptably smooth. 

The WebAssembly and PWA installs, will be great for teachers working inside school networks. But it needs to keep as much of the app feel as possible.

The multithreading in Uno seems to use multiple dotnet.native.workers. It hasn't given any obvious problems yet, outside of interop yet, but it's still being tested. Unfortunately, many of the background tasks involve database access which does require interop. Add to that cryptography because it's not implemented yet on the .net side.

So basically, I was looking for a way for C# code to call the db/crypto functions (webworker safe APIs) transparently from whichever worker instance it finds itself in, except for maybe calling ImportAsync if that worker was not set up. But the only way is to pass them to the main thread.

**Comment by Fydar** [comment]
I'm interested in seeing what the Blazor team can do with multithreading internally. I'm using client-side Blazor components embedded into my server-rendered application; but when the UI being unresponsive for the first second that I'm loading the page is incredibly disruptive in the user flow.

With threading, I suspect the unresponsive UI could be fixed.

I would love to even allow a Blazor runtime to be executed in a web worker to allow for sharing state between multiple browser tabs so that startup costs are only encountered once.

**Comment by pavel-zhur** [comment]
> > @pavelsavara I'm definitely wishing you luck. But it seems we won't have it for another 2+ years, if I understand you correctly.
> 
> I understand that you are waiting with some specific use-case for this ? Could you please share more details why/how you would benefit from threads in browser ? Also anyone else reading this, I would like to hear about your use-cases.

In my case, in the app that I'm building, the UI freezes while any time consuming operation is taking place (like deserialization of 10MB of json, or interoping with JS - for instance, storing data in the local cache, which also includes serialization).

As a result, while the app is busy, it is unable to immediately react to the user interaction.

It seems I need to make blazor process only extremely small pieces of data or to offload heavy data processing to [other](https://github.com/LostBeard/SpawnDev.BlazorJS) libraries that use web workers. I wish it all just worked natively and blazor could utilize multiple cores of the client browser out of the box.

**Comment by rootflood** [comment]
> > @pavelsavara I'm definitely wishing you luck. But it seems we won't have it for another 2+ years, if I understand you correctly.
> 
> I understand that you are waiting with some specific use-case for this ? Could you please share more details why/how you would benefit from threads in browser ? Also anyone else reading this, I would like to hear about your use-cases.

Task.Wait() method in WASM is blocked without threading
in some of major cases end of some frameworks there is wait method and if task.Wait don't work all of that code should change to async

**Comment by pavelsavara** [comment]
> Task.Wait() method in WASM is blocked 

We plan to keep it throwing PNSE on the UI thread (and any thread with JS interop). 
Reasons is that when the thread is synchronously blocked or spin-blocked, it could not serve the browser event loop.
1) handle any UI event or rendering
2) handle any networking event (HTTP and WS)
3) new threads can't be created
4) broser dev tools are not working properly. VS debugger is probably hanging too.
and any of above could lead to deadlocks.

`Task.Wait()` would start working on any `new Thread` or `Task.Run` in the thread pool.
So you could move your synchronously blocking code out of UI thread and rest should be fine.

Is that acceptable trade-off ? If not, why not ? 
What scenarios would be prevented by moving to thread pool ?
What would be better design option ?

**Comment by Matheos96** [comment]
> > @pavelsavara I'm definitely wishing you luck. But it seems we won't have it for another 2+ years, if I understand you correctly.
> 
> I understand that you are waiting with some specific use-case for this ? Could you please share more details why/how you would benefit from threads in browser ? Also anyone else reading this, I would like to hear about your use-cases.

Here is our use-case:

We have a Client-side only Blazor WASM app. Our main component is a PCB 3D Board Viewer. We load the data from our database but the data itself is not that complex (source from CAD formats). It does not technically contain the 3D data we need to render. So we build the 3D models on demand, based on our own Data model which is being deserialized and interpreted, all client side. The amount of details on such a board is immense, meaning we will have to do A LOT of calculations. For big boards we may have waiting times for more than  4 minutes, simply waiting for the deserialization (not too bad) and generation of 3D vertices, uvs and so on. During this time, the browser is pretty much completely frozen, we have some Task.Delay(1)'s here and there to update a progress bar in between but often we get the "not responding" message anyway. This is obviously a horrible user experience. Our dream is to be able to offload this generation to separate threads, for example separate board layers don't depend on each other and can be built completely in parallel. Ideally we want to show something really quickly, and then in the background keep calculating stuff that become available once done. This is obviously nothing but a dream at this point, still being restricted to the UI thread (and potentially web workers which we will look into). The whole thing does not get any better from the fact that we depend on A LOT of JS interop too... But that is another story.

So that is our use-case. One could ask why we don't do server-side rendering, but the fact is that we don't really have a deployed production "web" version yet. Currently we are providing this as an extension of our legacy phat client (embedding the SPA in a CEF Sharp window). One of our selling points is also the fact that we "make everything happen" in the browser, without the need for a backend. We also offer a "from file" option which keeps all data in the browser, meaning we definately have to do the calculations in the wasm app.

**Comment by curiousdannii** [comment]
Copying my use case from #68162:

I have an app I'm trying to port, which calls a blocking function in a non-managed dll:

```
        [DllImport("Glk")]
        internal static extern void glk_select(ref Event ev);
```

To port it to WASM/JS it instead needs to call an async JS function (ignore that it no longer takes any arguments):

```
        [JSImport("glk_select", "main.js")]
        internal static partial Task glk_select();
```

I've made an interface that lets me link to either the DLL or the JSImport, but I just need some way to block while I wait for the promise to resolve. And that's pretty much it!

(I also tried making the app async, but it's a large messy legacy app (28k+ LOC) and I would end up needing convert nearly everything to async functions, so threads seems a better option.) (And async functions not having `ref` or `out` arguments and VB.net not having tuple destructuring makes it even grosser to try to convert.)

**Comment by iSeiryu** [comment]
Copy-pasting use-cases from here: https://www.reddit.com/r/Blazor/comments/199o6e6/why_is_multithreading_such_a_requested_feature/

- You want to spawn up background threads to perform tasks that aren't directly related to UI things that only occasionally come back to UI

- You have things that can run in the background to load into a static property to show when ready and you don't want to tie up UI thread time on processing it - you expect the UI to just show it when ready

- You have some legacy CS code you want to include in the app - that's the whole point right, you share code but oops this code has a sync method and your UI is now fucked

- You hope to dear god you can have something like promises that you get in JS

- Prevent all runtime exceptions relating to threading. I sure had them early days before I fully realised what I had bought into, back to the legacy code integration thing

- Just why don't we have multithreading in what is otherwise a multi-threaded environment ... except for Blazor

**Comment by pavelsavara** [comment]
>    You hope to dear god you can have something like promises that you get in JS

I'm not sure I follow, could you please elaborate @iSeiryu  ? 
You can marshal `Task`/`Promise` with `JSImport`.

**Comment by iSeiryu** [comment]
@pavelsavara those are not my words. That's just a list of use-cases I came across yesterday. Btw, someone else in that Reddit thread asked a similar question.

**Comment by dennis-garavsky** [comment]
>enable blocking Task.Wait and lock() like APIs from C# user code on all threads

We will also benefit from this enhancement at DevExpress in certain products, and our customers will for sure benefit in their projects (while migrating shared code to .NET 8 and Blazor). 

For instance, in our application framework DevExpress XAF (powered by ASP.NET Core Blazor), we researched ways to plug in a [middle tier service](https://docs.devexpress.com/eXpressAppFramework/404398/data-security-and-safety/security-system/non-xaf/connect-to-the-efcore-middle-tier-service-from-non-xaf-applications) into our existing code base with Blazor WebAssembly and EF Core. The following code throws "_Cannot wait on monitors on this runtime_" with Blazor .NET 8 as it did 5 years ago when we started supporting Blazor in XAF and experimented with WebAssembly back then (dropped WebAssembly due to low performance and the lack of multi-threading support). Today we still rely on Blazor Server, but want to support WebAssembly render modes in the future.

```
public class WebApiSecuredDataServerClient
        protected override TResult Invoke<TResult>(string action) {
            return StaSafeHelper.Invoke(() => InvokeAsync<TResult>(action, CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult());
        }
```
As other readers also noted, we also want to reuse a large amount of our existing framework core code in Blazor WebAssembly, as we successfully did in Blazor Server, WinForms and WebForms for over 15 years now. Converting "nearly everything to async functions" in the framework core code is NOT realistic due to the huge breaking changes for other customers and rewrite costs, of course (it may be easier to drop everything and start over with a new product then).

Thank you for your consideration.

**Comment by pavelsavara** [comment]
> > enable blocking Task.Wait and lock() like APIs from C# user code on all threads

We are still re-considering this one. See my [other comment](https://github.com/dotnet/runtime/issues/85592#issuecomment-1904777373).

> Converting "nearly everything to async functions" in the framework core code is NOT realistic 

We understand that ^^.

Alternatives are
- if you could offload such code out of UI thread with moderate effort, this is best.
- multi-threading runtime will not allow any managed code in physical UI thread and run in in WebWorker (deputy). It will marshal all JS interop across physical threads. This is still work in progress POC, we hope to know more soon.
- allow you to take the risk (via feature flag) and allow you to spin-wait in UI thread and face deadlocks. We will not support this for production.

**Comment by curiousdannii** [comment]
The mono runtime is compiled using Emscripten. Is this also the case with AOT compilation? If that's the case, any chance that we could run ASYNCIFY like we would for any other Emscripten app? That could be another way of converting sync functions into async.

**Comment by glutio** [comment]
Hi, just trying to understand what is currently (.net8) supported and how it's going to change in .net9, it's been difficult to find definitive information. I am using dotnet webassembly without Blazor (based on wasm-experiments wasmbrowser template). From my experiments on .net8 I bellieve that if dotnet is running on the **main** thread then starting a new Thread or running a Task both **work**, but if dotnet is running in a **[WebWorker](https://github.com/ilonatommy/reactWithDotnetOnWebWorker?tab=readme-ov-file)** then it fails on startup (when `WasmEnableThreads` is true). Also `HttpClient` **does not work** from a **non-main** Thread or Task but it does work from the WebWorker's thread. I also tried .net9 preview but had some issues with it, one thing I recall is that it complained when I had JS->.NET->JS callstack, it said something like calling a synchronous JS method from a synchronous .NET method is not supported, while it works fine in .net8.

Could you please comment on it not being able to start Threads/Tasks from inside a Thread/Task or when dotnet is running in a WebWorker, and also on HttpClient not working inside Threads/Tasks? In my scenario I need to block execution waiting for user input via DOM elements, async in not an option in my case so I run .net outside the main thread. I tried different approaches and running dotnet in a WebWorker looks promising because I can block execution and use HttpClient but I cannot start new Tasks/Threads, is this going to be changed in .net9 or is it possible to do something about it in .net8? Thank you!

**Comment by glutio** [comment]
Trying .net9 preview 2, the all synchronous call .net->js->.net hangs the browser window. Seems like a regression from .net8

**Comment by glutio** [comment]
Thanks, I'm building a C# online playground that I want to be 100% clientside and suppot as many features as possible like blocking calls (Console.ReadLine), async programming with Tasks and multithreading with Threads. With the dotnet wasm MT it seems it could be possible. I'm using Roslyn to compile user code clientside and I want to run the user code in a separate thread so that if there's some bad logic like an infinite loop it wouldn't hang the UI and could be restarted without closing the browser window. What are the limitations of Threads vs Tasks and of running dotnet on the main thread vs in a [WebWorker](https://github.com/ilonatommy/reactWithDotnetOnWebWorker?tab=readme-ov-file)? E.g. as of preview2 HttpClient works in Tasks but not in Threads. Would you recommend one approach over the other for running the sandbox and usercode? Thanks!

**Comment by glutio** [comment]
Thanks. BTW, the default wasmbrowser template is not working with .net9-preview2, due to the above, saying "Cannot call synchronous JS function from inside a synchronous call to a C# method". If you remove the synchronous call it works but crashes after executing .net Main() (oddly I think it worked earlier today?). 
```
worker.onerror		@	dotnet.native.9.0.0-…8.5.kpnkj0mqsi.js:8
error (async)		
(anonymous)		@	dotnet.native.9.0.0-…8.5.kpnkj0mqsi.js:8
loadWasmModuleToWorker	@	dotnet.native.9.0.0-…8.5.kpnkj0mqsi.js:8
e.loadWasmModuleToWorker@	ui-thread.ts:243
Bn			@	ui-thread.ts:359
(anonymous)		@	ui-thread.ts:169
(anonymous)		@	startup.ts:247
```

**Comment by maxkatz6** [comment]
@pavelsavara would it be possible to call a sync .NET exported function (or a JSImport callback) without blocking JS thread? I.e. discarding a result. Making .NET function async is a decent workaround here, but I don't see why it should be forced for "fire and forget" scenarios. Thanks.

**Comment by maxkatz6** [comment]
Also, will new behavior be toggleable? It's not yet clear if we can support new threading model in .NET 9 WASM, since it has to work with sync requestAnimationFrame callbacks, webGL and Skia on top of Emscripten, something that yet needs to be tested.
But .NET 8 WASM threading model might be sufficient for most Avalonia apps (although we would love to make new one supported too).

**Comment by kekekeks** [comment]
I believe that the lack of access to the "true" UI thread would completely break WebGL interop. 
WebGL doesn't have a dedicated present/swapBuffers call and instead just presents render results once user code exit the current event callback. If JS API calls are getting enqueued from a worker thread, that would mean that webgl content will be "presented" on every single call, thus making any kind of rendering impossible.

Another problem would be WebGL usage with SkiaSharp: Skia uses WebGL internally and expect the calls to happen on the "true" UI thread rather than on a web worker.

**Comment by pavelsavara** [comment]
> "fire and forget" scenarios. 

I created `DiscardNoWait` for it and it's applied to `void` methods.
But it's not on public API, yet.
It executes asynchronously similar to methods returning `Task`, but it doesn't need to marshal the `Task`.
If you search this repo you can see it used in some tests. 
https://github.com/dotnet/runtime/blob/5111fdc0dc464f01647d6b6078342f451bf3a499/src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSMarshalerType.cs#L47-L54

You can help me by pushing it thru API review process (and have discussion about it's name on that PR).

> Could it be possibly to initialize .NET thread on a physical UI thread?

We will not support running managed code on UI thread, I'm confident now. 
It leads to too many deadlocks which are not possible to predict or even test against.

At the moment the UI thread is still attached to Mono VM, but that's going to change as soon as possible.

> Also, will new behavior be toggleable? 

There is toggle for allowing synchronous JSExport (with all the consequences of spin-blocking UI thread) 
We just have to make it into [MSBuild property](https://github.com/dotnet/runtime/issues/98652)
It's quite easy to cause deadlock with it, but not so bad as running managed code on UI thread.

> But .NET 8 WASM threading model might be sufficient for most Avalonia apps

Net8 WASM threading is very broken, my 2c: don't do it to your users.

> I believe that the lack of access to the "true" UI thread would completely break WebGL interop. 

I would like to learn more, how to make more native scenarios possible with MT.
This is not right place for detailed discussion, so I made https://github.com/dotnet/runtime/issues/101421 for it

**Comment by maxkatz6** [comment]
> You can help me by pushing it thru API review process

How can I help with the review process? Naming looks fine, easier to understand than "OneWay".

**Comment by pavelsavara** [comment]
> > You can help me by pushing it thru API review process
> 
> How can I help with the review process? Naming looks fine, easier to understand than "OneWay".

You can follow https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md 
Create issue with proposal coach it thru the API review meeting.
Also create PR which removes the API from https://github.com/dotnet/runtime/blob/bc9fc5a774d96f95abe0ea5c90fac48b38ed2e67/src/libraries/System.Runtime.InteropServices.JavaScript/src/CompatibilitySuppressions.xml#L15-L20

Which will make it public.

Examples are https://github.com/dotnet/runtime/issues?q=is%3Aissue+label%3Aapi-approved+is%3Aclosed

@maxkatz6

**Comment by kekekeks** [comment]
Will async callbacks be only supported for JSExport? We don't use those and are using `[JSMarshalAs(JSType.Function)]`.

If we try to change JS->.NET callbacks to return tasks, the SDK complains about `18>TimerHelper.cs(14,113): Error SYSLIB1072 : The type 'System.Func<System.Threading.Tasks.Task>' is not supported by source-generated JavaScript interop. The generated source will not handle marshalling of parameter 'callback'. For more information see https://aka.ms/dotnet-wasm-jsinterop (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1072)
`

What is the new intended way to pass delegates to JS side?

**Comment by pavelsavara** [comment]
> The type 'System.Func<System.Threading.Tasks.Task>' is not supported

That's a (known) gap. Could you please open separate issue for it and ping me there ? @kekekeks

**Comment by pavelsavara** [comment]
I'm closing this as complete for the scope of Net9 for the dotnet runtime.

### Summary

The difference between Net8 and Net9 threading is that it actually works now. 😅 
All the managed code is running in the web workers and that makes it possible to use blocking `.Wait` on all managed threads.
Limitation (by design) is that you can only call asynchronous C# methods as `JSExport` from the UI thread (main JS with DOM).
All JavaScript interop via `JSImport` is dispatched on to UI thread via low level async message. 
Another limitation by design is that you can't make nested synchronous callback from inside synchronous `JSImport` call.
HTTP and WebSocket connections are made from UI thread.

Status is still: experimental

### Blazor

For Blazor WASM, the multi-threading feature was [cut for Net9](https://github.com/dotnet/aspnetcore/issues/17730#issuecomment-2059602250). 
The biggest gaps in Blazor are described in https://github.com/dotnet/aspnetcore/issues/54365

### Demo

I updated the Raytracer demo https://pavelsavara.github.io/dotnet-wasm-raytracer/

### Performance

We didn't optimize for multi-threaded performance yet. 
As compared to single-threaded dotnet on WASM, it will be slightly slower per core, but you can use many cores now. 
It should also allow you to offload CPU intensive work from the main thread and keep the DOM responsive.

### Future work

There is draft of `JSWebWorker` API which we will explore in the future. It can allow you to interact with another JS than the UI thread. 
We didn't implement any of the sync-over-async scenarios in this release.

If there are bugs or missing features, **please create new issue on runtime repo** and ping me, we will triage and prioritize.

**Comment by pavelsavara** [comment]
The demo above is on Preview 3.
Net9 Preview 3 has most of it and Preview 4 has fixes and some cleanup, I expect that we will improve quality further.
If you run into issues you can always try [nightly build](https://aka.ms/dotnet/9.0.1xx/daily/dotnet-sdk-win-x64.zip) and see if it was already fixed.

**Comment by pavelsavara** [comment]
> Is any kind of debugging supposed to work with MT mode? I can't get vscode nor `/_framework/debug` to stop on breakpoints.

There is open issue for that https://github.com/dotnet/runtime/issues/81282 and few more for broken tests
